consensus/exit_probability_validator/exit_probability_validator.c

```diff
@@ -41,47 +63,132 @@ retcode_t iota_consensus_exit_prob_transaction_validator_is_valid(
+  if (ret != RC_OK) {
+    *is_valid = false;
+    return ret;
+  }
+  if (!*is_valid) {
+    log_error(WALKER_VALIDATOR_LOGGER_ID,
+              "Validation failed, tail is inconsistent\n");
     return RC_OK;
   }
 
+  bool below_max_depth = false;
+
+  // TODO - lowest_allowed_depth
+  ret = iota_consensus_exit_prob_transaction_validator_below_max_depth(
+      epv, tail_hash, epv->max_depth, &below_max_depth);
+  if (ret != RC_OK) {
+    return ret;
+  }
+
+  if (below_max_depth) {
+    *is_valid = false;
+    log_error(WALKER_VALIDATOR_LOGGER_ID,
+              "Validation failed, tail is below max depth\n");
+    return RC_OK;
+  }
+
+  // TODO - change schema and update solidity
+  /*TransactionViewModel transactionViewModel =
+  TransactionViewModel.fromHash(tangle, transactionHash); if
+  if (!transactionViewModel.isSolid()) {
+    log.debug("Validation failed: {} is not solid", transactionHash);
+    return false;
+  }*/
+
   *is_valid = true;
 
   return ret;
 }
 
 retcode_t iota_consensus_exit_prob_transaction_validator_below_max_depth(
-    const exit_prob_transaction_validator_t *wv, trit_array_p tip,
-    size_t depth) {
-  return RC_CONSENSUS_NOT_IMPLEMENTED;
+    exit_prob_transaction_validator_t *epv, trit_array_p tail_hash,
+    uint32_t lowest_allowed_depth, bool *below_max_depth) {
+  retcode_t res = RC_OK;
+  flex_trit_t curr_hash_trits[FLEX_TRIT_SIZE_243];
+
+  hash_queue_t non_analyzed_hashes = NULL;
+  hash_queue_push(&non_analyzed_hashes, tail_hash->trits);
+
+  hash_set_t visited_hashes = NULL;
+  while (non_analyzed_hashes != NULL) {
+    if (hash_set_size(&visited_hashes) == epv->max_analyzed_txs) {
+      log_error(WALKER_VALIDATOR_LOGGER_ID,
+                "Validation failed, exceeded num of transactions\n");
+      *below_max_depth = false;
+      break;
+    }
+
+    memcpy(curr_hash_trits, hash_queue_peek(non_analyzed_hashes),
+           FLEX_TRIT_SIZE_243);
+    hash_queue_pop(&non_analyzed_hashes);
+
+    if (hash_set_contains(&visited_hashes, curr_hash_trits)) {
+      continue;
+    }
+
+    // Mark the transaction as visited
+    if ((res = hash_set_add(&visited_hashes, curr_hash_trits))) {
+      break;
+    }
+
+    // Load the transaction
+    struct _iota_transaction curr_tx_s;
+    iota_transaction_t curr_tx = &curr_tx_s;
+
+    iota_stor_pack_t pack = {.models = (void **)(&curr_tx),
+                             .capacity = 1,
+                             .num_loaded = 0,
+                             .insufficient_capacity = false};
+
+    TRIT_ARRAY_DECLARE(hash_trits_array, NUM_TRITS_HASH);
+    memcpy(hash_trits_array.trits, curr_hash_trits, FLEX_TRIT_SIZE_243);
+    res = iota_tangle_transaction_load(epv->tangle, TRANSACTION_COL_HASH,
+                                       &hash_trits_array, &pack);
+    curr_tx = (iota_transaction_t)pack.models[0];
+    bool tail_is_not_genesis =
+        (curr_tx->snapshot_index != 0 ||
+         memcmp(null_hash, curr_tx->hash, FLEX_TRIT_SIZE_243) == 0);
```

Comment:
Maybe I'm missing something but the variable is called `tail_is_not_genesis` and it's true if `memcp() == 0` i.e is genesis ?

Reply:
but than it's hash must not be just `9999...`